near-contract-standards/src/fungible_token/implementation.rs

```diff
@@ -0,0 +1,256 @@
+use crate::fungible_token::core::FungibleTokenCore;
+use crate::fungible_token::resolver::FungibleTokenResolver;
+use crate::storage_manager::{AccountStorageBalance, StorageManager};
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::collections::LookupMap;
+use near_sdk::json_types::{ValidAccountId, U128};
+use near_sdk::{
+    assert_one_yocto, env, ext_contract, log, AccountId, Balance, Gas, Promise, PromiseResult,
+    StorageUsage,
+};
+
+const GAS_FOR_RESOLVE_TRANSFER: Gas = 5_000_000_000_000;
+const GAS_FOR_FT_TRANSFER_CALL: Gas = 25_000_000_000_000 + GAS_FOR_RESOLVE_TRANSFER;
+
+const NO_DEPOSIT: Balance = 0;
+
+#[ext_contract(ext_self)]
+trait FungibleTokenResolver {
+    fn ft_resolve_transfer(
+        &mut self,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+    ) -> U128;
+}
+
+#[ext_contract(ext_fungible_token_receiver)]
+pub trait FungibleTokenReceiver {
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: U128, msg: String) -> Promise;
+}
+
+/// Implementation of a FungibleToken standard.
+/// Allows to include NEP-141 compatible token to any contract.
+/// There are next traits that any contract must implement:
+///     - FungibleTokenCore -- interface with ft_transfer methods. FungibleToken provides methods for it.
+///     - FungibleTokenMetaData -- return metadata for the token in NEP-148, up to contract to implement.
+///     - StorageManager -- inteface for NEP-145 for allocating storage per account. FungibleToken provides methods for it.
+///
+/// For example usage, see examples/fungible-token/src/lib.rs.
+/// ```
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FungibleToken {
+    /// AccountID -> Account balance.
+    pub accounts: LookupMap<AccountId, Balance>,
+
+    /// Total supply of the all token.
+    pub total_supply: Balance,
+
+    /// The storage size in bytes for one account.
+    pub account_storage_usage: StorageUsage,
+}
+
+impl FungibleToken {
+    pub fn new(prefix: &[u8]) -> Self {
+        let mut this = Self {
+            accounts: LookupMap::new(prefix.to_vec()),
+            total_supply: 0,
+            account_storage_usage: 0,
+        };
+        let initial_storage_usage = env::storage_usage();
```

Comment:
Does the line above add anything to storage?  If not, perhaps the lines should be switched to be explicit that it's the storage used before running the constructor.

Reply:
`env::storage_usage()` returns the value in bytes that is immediately updates when we modify internal value for inserting the temp account into the structure. We should probably refactor it a little to create a method that explains the logic below.